frontend/test/async_tests/test_async.py

```diff
@@ -0,0 +1,373 @@
+        def circuit2(params, x):
+            qml.RY(params[0], wires=0)
+            qml.RZ(params[1], wires=1)
+            qml.CNOT(wires=[x, 1])
+            return qml.expval(qml.PauliX(wires=0))
+
+        @qml.qnode(device=dev)
+        def circuit3(params, x):
+            qml.RZ(params[0], wires=0)
+            qml.RX(params[1], wires=1)
+            qml.CNOT(wires=[x, 1])
+            return qml.expval(qml.PauliY(wires=0))
+
+        new_params = jnp.array([circuit1(params, x), circuit2(params, y)])
+        return circuit3(new_params, z)
+
+    params = jnp.array([1.0, 2.0])
+    msg = "Unrecoverable error"
+    with pytest.raises(RuntimeError, match=msg):
+        qjit(async_qnodes=True)(multiple_qnodes)(params)
+
+
+# TODO: add the following diff_methods once issue #419 is fixed:
+# ("parameter-shift", "auto"), ("adjoint", "auto")]
+@pytest.mark.parametrize("diff_methods", [("finite-diff", "fd")])
+@pytest.mark.parametrize("inp", [(1.0)])
+def test_gradient_exception(inp, diff_methods, backend):
+    """Parameter shift and finite diff generate multiple QNode that are run async."""
+
+    def f(x, y):
+        qml.RX(x * 2, wires=0)
+        qml.CNOT(wires=[0, y])
+        return qml.expval(qml.PauliY(0))
+
+    @qjit(async_qnodes=True)
+    def compiled(x: float):
+        g = qml.qnode(qml.device(backend, wires=1), diff_method=diff_methods[0])(f)
+        h = grad(g, method=diff_methods[1], argnum=[0])
+        return h(x, 0)
+
+    msg = "Unrecoverable error"
+    with pytest.raises(RuntimeError, match=msg):
+        compiled(inp)
+
+
+def test_exception_in_loop(backend):
+    "Test exception happening in a loop."
+
+    @qjit(async_qnodes=True)
+    @qml.qnode(qml.device(backend, wires=3))
+    def circuit(n):
+        @while_loop(lambda v: v[0] < v[1])
+        def loop(v):
+            qml.CNOT(wires=[n, v[0]])
+            return v[0] + 1, v[1]
+
+        loop((0, 3))
+        return measure(wires=0)
+
+    msg = "Unrecoverable error"
+    # Exception in beginning
+    with pytest.raises(RuntimeError, match=msg):
+        circuit(0)
+    # Exception in middle
+    with pytest.raises(RuntimeError, match=msg):
+        circuit(1)
+    # Exception in end
+    with pytest.raises(RuntimeError, match=msg):
+        circuit(2)
+
+
+def test_exception_in_loop2(backend):
+    "Test exception happening in a loop while one qnode succeeds."
+
+    @qml.qnode(qml.device(backend, wires=3))
+    def bad(n):
+        @while_loop(lambda v: v[0] < v[1])
+        def loop(v):
+            qml.CNOT(wires=[n, v[0]])
+            return v[0] + 1, v[1]
+
+        loop((0, 3))
+        return measure(wires=0)
+
+    @qml.qnode(qml.device(backend, wires=3))
+    def good():
+        return qml.state()
+
+    @qjit(async_qnodes=True)
+    def wrapper(n):
+        x = good()
+        y = bad(n)
+        return x + y
+
+    msg = "Unrecoverable error"
+    # Exception in beginning
+    with pytest.raises(RuntimeError, match=msg):
+        wrapper(0)
+
```

Comment:
can we add checks within async execution of multiple qnodes in calls to `catalyst.adjoint` and `catalyst.ctrl`?

Reply:
I'll add it, but it shouldn't matter which operation is in the program. The exception handling code happens in the caller of the qnode.

Reply:
Here's adjoint: https://github.com/PennyLaneAI/catalyst/pull/447/commits/402a4555c3f6cb30e7cbcdf62b711468a1048d43